### PR TITLE
Feature constraints

### PR DESCRIPTION
**Pull request by ronanstokes-db**
## Proposed changes

Added support for constraints

## Types of changes

What types of changes does your code introduce to dbldatagen?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Change to tutorials, tests or examples
- [x] Non code change (readme, images or other non-code assets)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. 
If you're unsure about any of them, don't hesitate to ask. We're here to help! 
This is simply a reminder of what we are going to look for before merging your code._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Submission does not reduce code coverage numbers
- [x] Submission does not increase alerts or messages from prospector / lint

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you 
did and what alternatives you considered, etc...

### CODE REVIEW

**Review by howardwu-db — CHANGES_REQUESTED** [review]
Probably add comments to the abstract classes instead of using ABC

File: dbldatagen/column_spec_options.py [hunk 1]
```diff
@@ -47,6 +47,9 @@ class ColumnSpecOptions(object):
     :param baseColumn: Either the string name of the base column, or a list of columns to use to
                         control data generation. The option ``baseColumns`` is an alias for ``baseColumn``.
 
+    :param baseColumnType: Determines how the value is derived from the base column. Possible values are 'auto',
```
[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/column_spec_options.py:50]: Fixed in coming checkin

[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/column_spec_options.py]: moved to different PR

File: dbldatagen/constraints/constraint.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        elif relation == "<=":
+            return column <= valueExpression
+        elif relation in ["!=", "<>"]:
+            return column != valueExpression
+        elif relation in ["=", "=="]:
+            return column == valueExpression
+        else:
+            raise ValueError(f"Unsupported relation type '{relation}")
+
+    @classmethod
```
[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/constraints/constraint.py]: The class has been reworked so no longer applicable (i think)

**Review by howardwu-db — APPROVED** [review]

**Pull request merged**